### ISSUE TITLE
Orion 3.1.0 crashes with PEP-Proxy (Wilma) 8.0.0

### ISSUE THREAD

**Issue by jason-fox** [telefonicaid/fiware-orion#3884]
Consider a set-up with a [PEP Proxy](https://github.com/ging/fiware-pep-proxy) before an Orion Instance - e.g. this [docker-compose.yml](https://github.com/FIWARE/tutorials.PEP-Proxy/blob/master/docker-compose/orion.yml) 

The PEP Proxy annotates the forwarded request with extra headers. With Orion 3.0.0 and earlier this is no problem, however the following request (either directly using Curl or indirectly using the tutorial) causes the 3.1.0 Broker to crash:

```console
curl -L -X PATCH 'http://localhost:1026/v2/entities/Bell:001/attrs' \
-H 'fiware-service: openiot' \
-H 'fiware-servicepath: /' \
-H 'Content-Type: application/json' \
-H 'host: orion-proxy:1027' \
-H 'accept: application/json' \
-H 'connection: close' \
-H 'X-Nick-Name: cccccccc-good-0000-0000-000000000000' \
-H 'x-auth-token: 2ca8a58ec971bfc6b0bcda8443931883eb5900d5' \
-H 'X-Display-Name: ' \
-H 'X-Roles: [{\"id\":\"security-role-0000-0000-000000000000\",\"name\":\"Security Team\"}]' \
-H 'X-Organizations: []' \
-H 'X-Eidas-Profile: {}' \
-H 'X-App-Id: tutorial-dckr-site-0000-xpresswebapp' \
-H 'x-forwarded-for: ::ffff:172.18.1.7' \
--data-raw '{
  "ring": {
      "type" : "command",
      "value" : ""
  }
}'
```

Specifically it is the `x-forwarded-for` header that can no longer be processed - the command works with Orion 3.0.0 or if the `x-forwarded-for` header is left out of the request.

The  log is shown below ( The `ring` attribute is set up as a Command via a registration.)

```text
time=2021-06-11T17:28:22.372Z | lvl=ERROR | corr=6bb56dd4-cada-11eb-a37b-0242ac120109 | trans=1623432459-057-00000000046 | from=::ffff:172.18.1./ | srv=openiot | subsrv=/ | comp=Orion | op=postUpdateContext.cpp[642]:postUpdateContext | msg=Internal Error (attribute 'ring' not found)
time=2021-06-11T17:28:22.372Z | lvl=INFO | corr=6bb56dd4-cada-11eb-a37b-0242ac120109 | trans=1623432459-057-00000000046 | from=::ffff:172.18.1./ | srv=openiot | subsrv=/ | comp=Orion | op=logTracing.cpp[146]:logInfoFwdStart | msg=Starting forwarding for PATCH /v2/entities/Bell:001/attrs
time=2021-06-11T17:28:26.388Z | lvl=INFO | corr=6e1b02c8-cada-11eb-89b0-0242ac120109 | trans=1623432459-057-00000000048 | from=172.18.1.9 | srv=<none> | subsrv=<none> | comp=Orion | op=logTracing.cpp[79]:logInfoRequestWithoutPayload | msg=Request received: GET /version, response code: 200
free(): invalid pointer
```

**Comment by fgalan** [comment]
Thanks for the report! I'm trying to reproduce the case, but to do it precisely I'd need also the curl commands corresponding to:

* Create the "Bell:001" entity
* Create the registration for the ring command (hint: if the registration has been done by a IOTA, `GET /v2/registrations` will provide clues on how to reproduce the same using curl)

@jason-fox could you provide also that information, please?

In addition, I have one doubt... Looking to the log fragment it seems the crash takes place when a `GET /version` request is done _after_ the PATCH shown in the curl command, but not when processing PATCH itself. It's my understanding correct?

**Comment by jason-fox** [comment]
You can provision the bell device with the following script: https://github.com/FIWARE/tutorials.PEP-Proxy/blob/master/provision-devices

You can reproduce the error by amending [this](https://github.com/FIWARE/tutorials.PEP-Proxy/blob/master/.env#L3) to `3.1.0` and running `./services start` followed by the curl statement defined above.

The `/version` is the heartbeat healthcheck - I don't think that is causing the crash, but it may have run at about the same time.

**Comment by fgalan** [comment]
> 
> 
> You can provision the bell device with the following script: https://github.com/FIWARE/tutorials.PEP-Proxy/blob/master/provision-devices
> 
> You can reproduce the error by amending [this](https://github.com/FIWARE/tutorials.PEP-Proxy/blob/master/.env#L3) to `3.1.0` and running `./services start` followed by the curl statement defined above. In the actual tutorial when the app has a valid user, it sends the statement to the PEP-Proxy which forwards to Orion. Without the `x-forwarded-for` header annotation there is no crash. The PEP proxy is not needed since I can cause the crash by mimicking the forwarded request directly.

I'm trying to reproduce the bug using just Orion and curl, without PEP, IOTAs, docker, etc. That way is much easier to debug.

Could you provide at least the result of `GET /v2/registrations` in your running system? With that information, I think I could craft the equivalent curl for the registration.

> The `/version` is the heartbeat healthcheck - I don't think that is causing the crash, but it may have run at about the same time.

Looking to log traces timestamp, around 4 seconds has passed between PATCH and GET /version resquests. It seems too much time to be related...

Another useful piece of information to debug I would need, please: the contextBroker CLI and env vars. This is typically printed at first traces in logs just after starting Orion, eg:

```
time=2021-06-15T12:13:02.370Z | lvl=INFO | corr=N/A | trans=N/A | from=N/A | srv=N/A | subsrv=N/A | comp=Orion | op=contextBroker.cpp[1063]:main | msg=start command line <contextBroker -fg -logLevel INFO>
time=2021-06-15T12:13:02.370Z | lvl=INFO | corr=N/A | trans=N/A | from=N/A | srv=N/A | subsrv=N/A | comp=Orion | op=contextBroker.cpp[937]:logEnvVars | msg=env var ORION_PORT (-port): 1026
time=2021-06-15T12:13:02.370Z | lvl=INFO | corr=N/A | trans=N/A | from=N/A | srv=N/A | subsrv=N/A | comp=Orion | op=contextBroker.cpp[1137]:main | msg=Orion Context Broker is running
```

**Comment by jason-fox** [comment]
```json
[
    {
        "id": "60c89c3733e08040363dfb0b",
        "dataProvided": {
            "entities": [
                {
                    "id": "Bell:001",
                    "type": "Bell"
                }
            ],
            "attrs": [
                "ring"
            ]
        },
        "provider": {
            "http": {
                "url": "http://iot-agent:4041"
            },
            "supportedForwardingMode": "all",
            "legacyForwarding": true
        },
        "status": "active"
    }
]
```

### Start up

```text
time=2021-06-15T12:24:45.613Z | lvl=INFO | corr=N/A | trans=N/A | from=N/A | srv=N/A | subsrv=N/A | comp=Orion | op=contextBroker.cpp[1063]:main | msg=start command line </usr/bin/contextBroker -fg -multiservice -ngsiv1Autocast -disableFileLog -dbhost mongo-db -logLevel DEBUG>
time=2021-06-15T12:24:45.617Z | lvl=INFO | corr=N/A | trans=N/A | from=N/A | srv=N/A | subsrv=N/A | comp=Orion | op=contextBroker.cpp[1137]:main | msg=Orion Context Broker is running
time=2021-06-15T12:24:45.806Z | lvl=INFO | corr=N/A | trans=N/A | from=N/A | srv=N/A | subsrv=N/A | comp=Orion | op=mongoConnectionPool.cpp[506]:mongoConnectionPoolInit | msg=Connected to mongodb://mongo-db/?connectTimeoutMS=10000 (dbName: orion, poolsize: 10)
time=2021-06-15T12:24:45.814Z | lvl=INFO | corr=N/A | trans=N/A | from=N/A | srv=N/A | subsrv=N/A | comp=Orion | op=contextBroker.cpp[1263]:main | msg=Startup completed
```

### Registration

```text
time=2021-06-15T12:25:27.284Z | lvl=INFO | corr=c4235420-cdd4-11eb-871f-0242ac120109 | trans=1623759885-620-00000000008 | from=172.18.1.3 | srv=openiot | subsrv=/ | comp=Orion | op=logTracing.cpp[130]:logInfoRequestWithPayload | msg=Request received: POST /v2/registrations, request payload (132 bytes): {"dataProvided":{"entities":[{"type":"Bell","id":"Bell:001"}],"attrs":["ring"]},"provider":{"http":{"url":"http://iot-agent:4041"}}}, response code: 201
```

### Failed Command

```text
time=2021-06-15T12:25:27.404Z | lvl=INFO | corr=71de7172-afe6-4833-8623-f68d1426a92c | 
trans=1623759885-620-00000000009 | from=172.18.1.3 | srv=openiot | subsrv=/ | comp=Orion | op=logTracing.cpp[130]:logInfoRequestWithPayload | msg=Request received: POST /v2/entities?options=upsert, request payload (477 bytes): {"id":"Bell:001","type":"Bell","refStore":{"type":"Relationship","value":"urn:ngsi-ld:Store:001"},"category":{"type":"Text","value":["actuator"]},"controlledProperty":{"type":"Text","value":"noiseLevel"},"function":{"type":"Text","value":["onOff"]},"supportedProtocol":{"type":"Text","value":["ul20"]},"ring_status":{"type":"commandStatus","value":"UNKNOWN"},"ring_info":{"type":"commandResult","value":" "},"TimeInstant":{"type":"DateTime","value":"2021-06-15T12:25:27.321Z"}}, response code: 204
time=2021-06-15T12:26:35.109Z | lvl=INFO | corr=ec9827d2-cdd4-11eb-8643-0242ac120109 | trans=1623759885-620-00000000026 | from=::ffff:127.0.0.1<none> | srv=<none> | subsrv=<none> | 
time=2021-06-15T12:27:34.211Z | lvl=ERROR | corr=0fd1c302-cdd5-11eb-8465-0242ac120109 | trans=1623759885-620-00000000039 | from=::ffff:172.18.1./ | srv=openiot | subsrv=/ | comp=Orion | op=postUpdateContext.cpp[642]:postUpdateContext | msg=Internal Error (attribute 'ring' not found)
time=2021-06-15T12:27:34.212Z | lvl=INFO | corr=0fd1c302-cdd5-11eb-8465-0242ac120109 | trans=1623759885-620-00000000039 | from=::ffff:172.18.1./ | srv=openiot | subsrv=/ | comp=Orion | op=logTracing.cpp[146]:logInfoFwdStart | msg=Starting forwarding for PATCH /v2/entities/Bell:001/attrs
time=2021-06-15T12:27:35.771Z | lvl=INFO | corr=10c0487e-cdd5-11eb-8120-0242ac120109 | trans=1623759885-620-00000000041 | from=172.18.1.9 | srv=<none> | subsrv=<none> | comp=Orion | op=logTracing.cpp[79]:logInfoRequestWithoutPayload | msg=Request received: GET /version, response code: 200
free(): invalid pointer
```

There was no `/version` this time.

**Comment by jason-fox** [comment]
Next run:

```text
time=2021-06-15T12:38:19.987Z | lvl=INFO | corr=90bc074c-cdd6-11eb-a181-0242ac120109 | trans=1623760540-821-00000000016 | from=::ffff:127.0.0.1<none> | srv=<none> | subsrv=<none> | comp=Orion | op=logTracing.cpp[79]:logInfoRequestWithoutPayload | msg=Request received: GET /version, response code: 200
time=2021-06-15T12:38:22.328Z | lvl=ERROR | corr=922047ec-cdd6-11eb-abb9-0242ac120109 | trans=1623760540-821-00000000017 | from=::ffff:172.18.1./ | srv=openiot | subsrv=/ | comp=Orion | op=postUpdateContext.cpp[642]:postUpdateContext | msg=Internal Error (attribute 'ring' not found)
time=2021-06-15T12:38:22.328Z | lvl=INFO | corr=922047ec-cdd6-11eb-abb9-0242ac120109 | trans=1623760540-821-00000000017 | from=::ffff:172.18.1./ | srv=openiot | subsrv=/ | comp=Orion | op=logTracing.cpp[146]:logInfoFwdStart | msg=Starting forwarding for PATCH /v2/entities/Bell:001/attrs
free(): invalid pointer
```

There was no `/version` this time.

**Comment by fgalan** [comment]
Thank you for so detailed report!

I'll have a look and comment in this issue.

**Comment by fgalan** [comment]
I have done the following test:

* Two Orion processes: one acting as CB, other acting as CPr (taking the role of the IOTA in your case)
* Database flushed before starting the tests:

```
use orion-openiot
db.dropDatabase()
use orioncpr-openiot
db.dropDatabase()
```

I start the two CB processes, this way (in separate terminals). Note that CB runs the same way you show in your report (except `-dbhost` that in my case is localhost).

```
contextBroker -fg -multiservice -ngsiv1Autocast -disableFileLog -dbhost localhost -logLevel DEBUG
contextBroker -fg -multiservice -ngsiv1Autocast -disableFileLog -dbhost localhost -logLevel DEBUG -db orioncpr -port 4041 -pidpath /tmp/orioncpr.pid
```

I check version, this way:

```
$ curl -s -S localhost:1026/version | grep version
  "version" : "3.1.0",
  "libversions": {
```

-----------------

Setup the scenario with:

Registration at CB, using the same payload shown in your report traces, except the URL (to point Orion as CPr in my case):

```
curl -vvvv localhost:1026/v2/registrations -s -S -H 'Content-Type: application/json' -H 'fiware-service: openiot' -H 'fiware-servicepath: /' -d @- <<EOF
{"dataProvided":{"entities":[{"type":"Bell","id":"Bell:001"}],"attrs":["ring"]},"provider":{"http":{"url":"http://localhost:4041/v2"}}}
EOF
```

showing this trace in CB logs:

```
time=2021-06-16T12:14:55.528Z | lvl=INFO | corr=75ffdfa2-ce9c-11eb-a327-000c29df7908 | trans=1623844833-131-00000000015 | from=0.0.0.0 | srv=openiot | subsrv=/ | comp=Orion | op=logTracing.cpp[130]:logInfoRequestWithPayload | msg=Request received: POST /v2/registrations, request payload (135 bytes): {"dataProvided":{"entities":[{"type":"Bell","id":"Bell:001"}],"attrs":["ring"]},"provider":{"http":{"url":"http://localhost:4041/v2"}}}, response code: 201
```

Entity creation at CB, using the same payload shown in your report traces:

```
curl -vvvv localhost:1026/v2/entities?options=upsert -s -S -H 'Content-Type: application/json' -H 'fiware-service: openiot' -H 'fiware-servicepath: /' -d @- <<EOF
{"id":"Bell:001","type":"Bell","refStore":{"type":"Relationship","value":"urn:ngsi-ld:Store:001"},"category":{"type":"Text","value":["actuator"]},"controlledProperty":{"type":"Text","value":"noiseLevel"},"function":{"type":"Text","value":["onOff"]},"supportedProtocol":{"type":"Text","value":["ul20"]},"ring_status":{"type":"commandStatus","value":"UNKNOWN"},"ring_info":{"type":"commandResult","value":" "},"TimeInstant":{"type":"DateTime","value":"2021-06-15T12:25:27.321Z"}}
EOF
```

showing this trace in CB logs:

```
time=2021-06-16T12:14:55.665Z | lvl=INFO | corr=760e762a-ce9c-11eb-ab96-000c29df7908 | trans=1623844833-131-00000000016 | from=0.0.0.0 | srv=openiot | subsrv=/ | comp=Orion | op=logTracing.cpp[130]:logInfoRequestWithPayload | msg=Request received: POST /v2/entities?options=upsert, request payload (477 bytes): {"id":"Bell:001","type":"Bell","refStore":{"type":"Relationship","value":"urn:ngsi-ld:Store:001"},"category":{"type":"Text","value":["actuator"]},"controlledProperty":{"type":"Text","value":"noiseLevel"},"function":{"type":"Text","value":["onOff"]},"supportedProtocol":{"type":"Text","value":["ul20"]},"ring_status":{"type":"commandStatus","value":"UNKNOWN"},"ring_info":{"type":"commandResult","value":" "},"TimeInstant":{"type":"DateTime","value":"2021-06-15T12:25:27.321Z"}}, response code: 204
```

Entity creation at CPr only for the attribute representing the command (`ring`):

```
curl -vvvv localhost:4041/v2/entities?options=upsert -s -S -H 'Content-Type: application/json' -H 'fiware-service: openiot' -H 'fiware-servicepath: /' -d @- <<EOF
{"id":"Bell:001","type":"Bell","ring":{"type":"command","value":"placeholder"}}
EOF
```

showing this trace in CPr logs:

```
time=2021-06-16T12:14:56.086Z | lvl=INFO | corr=7649be56-ce9c-11eb-a088-000c29df7908 | trans=1623844929-132-00000000005 | from=0.0.0.0 | srv=openiot | subsrv=/ | comp=Orion | op=logTracing.cpp[130]:logInfoRequestWithPayload | msg=Request received: POST /v2/entities?options=upsert, request payload (79 bytes): {"id":"Bell:001","type":"Bell","ring":{"type":"command","value":"placeholder"}}, response code: 204
```

---------------

Now I run the curl command in your initial post in this issue:

```
curl -L -X PATCH 'http://localhost:1026/v2/entities/Bell:001/attrs' \
-H 'fiware-service: openiot' \
-H 'fiware-servicepath: /' \
-H 'Content-Type: application/json' \
-H 'host: orion-proxy:1027' \
-H 'accept: application/json' \
-H 'connection: close' \
-H 'X-Nick-Name: cccccccc-good-0000-0000-000000000000' \
-H 'x-auth-token: 2ca8a58ec971bfc6b0bcda8443931883eb5900d5' \
-H 'X-Display-Name: ' \
-H 'X-Roles: [{\"id\":\"security-role-0000-0000-000000000000\",\"name\":\"Security Team\"}]' \
-H 'X-Organizations: []' \
-H 'X-Eidas-Profile: {}' \
-H 'X-App-Id: tutorial-dckr-site-0000-xpresswebapp' \
-H 'x-forwarded-for: ::ffff:172.18.1.7' \
--data-raw '{
  "ring": {
      "type" : "command",
      "value" : ""
  }
}'
```

I get in CB logs:

```
time=2021-06-16T12:15:31.574Z | lvl=ERROR | corr=8b8235aa-ce9c-11eb-b157-000c29df7908 | trans=1623844833-131-00000000017 | from=::ffff:172.18.1.0.0.0.0 | srv=openiot | subsrv=/ | comp=Orion | op=postUpdateContext.cpp[642]:postUpdateContext | msg=Internal Error (attribute 'ring' not found)
time=2021-06-16T12:15:31.574Z | lvl=INFO | corr=8b8235aa-ce9c-11eb-b157-000c29df7908 | trans=1623844833-131-00000000017 | from=::ffff:172.18.1.0.0.0.0 | srv=openiot | subsrv=/ | comp=Orion | op=logTracing.cpp[146]:logInfoFwdStart | msg=Starting forwarding for PATCH /v2/entities/Bell:001/attrs
time=2021-06-16T12:15:31.593Z | lvl=INFO | corr=8b8235aa-ce9c-11eb-b157-000c29df7908; cbfwd=1 | trans=1623844833-131-00000000018 | from=::ffff:172.18.1.0.0.0.0 | srv=openiot | subsrv=/ | comp=Orion | op=logTracing.cpp[212]:logInfoFwdRequest | msg=Request forwarded (regId: 60c9eb3f13deb14aac0e328e): POST http://localhost:4041/v2/op/update, request payload (119 bytes): {"entities":[{"id":"Bell:001","type":"Bell","ring":{"type":"command","value":"","metadata":{}}}],"actionType":"update"}, response payload (0 bytes): , response code: 204
time=2021-06-16T12:15:31.594Z | lvl=INFO | corr=8b8235aa-ce9c-11eb-b157-000c29df7908 | trans=1623844833-131-00000000017 | from=::ffff:172.18.1.0.0.0.0 | srv=openiot | subsrv=/ | comp=Orion | op=logTracing.cpp[130]:logInfoRequestWithPayload | msg=Request received: PATCH /v2/entities/Bell:001/attrs, request payload (64 bytes): {
  "ring": {
      "type" : "command",
      "value" : ""
  }
}, response code: 204
```

and the CB doesn't crash.

Note the first two lines above are the same as in your logs. The next lines corresponds to a successful forwarding case.

For completeness, this is the log I get in CPr:

```
time=2021-06-16T12:15:31.593Z | lvl=INFO | corr=8b8235aa-ce9c-11eb-b157-000c29df7908; cbfwd=1 | trans=1623844929-132-00000000006 | from=0.0.0.0 | srv=openiot | subsrv=/ | comp=Orion | op=logTracing.cpp[130]:logInfoRequestWithPayload | msg=Request received: POST /v2/op/update, request payload (119 bytes): {"entities":[{"id":"Bell:001","type":"Bell","ring":{"type":"command","value":"","metadata":{}}}],"actionType":"update"}, response code: 204
```

-------------

So, in sum, I haven't been able to reproduce the case... I guess that my test is not the same as yours. Let's try to find the differences in order to know what's going on.

Maybe the IOTA is your case not acting exactly as Orion-as-CPr in my case? Maybe the response send by the IOTA is wrong some way in your case and that's causing Orion to crash?

It would be great if you could get (from IOTA logs, I gest) the request/response it receives/sends when Orion forwards the update to it and add it as a comment to the issue, please.

**Comment by fgalan** [comment]
Another idea: get the `core` generated by CB upon crashing in your case, then using `gdb` to get the backtrace so we know the point in the code in which it's crashing.

**Comment by wistefan** [comment]
I was able to reproduce it exactly the same way as Jason did. I also used the docker-compose from the tutorials and the PATCH request, so its not a problem from Jasons machine. 
I can not find a suspicious log inside the IoT-Agent.

**Comment by jason-fox** [comment]
@fgalan I'm not sure how you're testing this but  maybe you're **not** getting a crash because you're running Orion on bare metal and we **are** getting a crash when containerized. So it **may** be an issue within the dockerization of Orion 3.1.0 - has it changed since 3.0.0?

**Comment by fgalan** [comment]
> @fgalan I'm not sure how you're testing this but maybe you're **not** getting a crash because you're running Orion on bare metal and we **are** getting a crash when containerized. So it **may** be an issue within the dockerization of Orion 3.1.0 - has it changed since 3.0.0?

Nothing relevant.

In https://github.com/telefonicaid/fiware-orion/compare/3.0.0...3.1.0, if you look for docker/Dockerfile y Files the only change between versions is to compile a different version of CMake due to build problems with the CentOS 8 new version with libmongoc library. But this should have any impact at all in Orion.

**Comment by fgalan** [comment]
> 
> 
> @fgalan I'm not sure how you're testing this but maybe you're **not** getting a crash because you're running Orion on bare metal and we **are** getting a crash when containerized. So it **may** be an issue within the dockerization of Orion 3.1.0 - has it changed since 3.0.0?

Maybe you could do a test on bare metal to confirm this hypothesis?

**Comment by fgalan** [comment]
> Another idea: get the `core` generated by CB upon crashing in your case, then using `gdb` to get the backtrace so we know the point in the code in which it's crashing.

Can you get this in your docker setup?

**Comment by fgalan** [comment]
Could you do the test using using `x-dedwarfor-for` (a header that is not processed by Orion but which name have the same size than `x-forwarded-for`) instead of `x-forwarded-for`?

This would be the curl:

```
curl -L -X PATCH 'http://localhost:1026/v2/entities/Bell:001/attrs' \
-H 'fiware-service: openiot' \
-H 'fiware-servicepath: /' \
-H 'Content-Type: application/json' \
-H 'host: orion-proxy:1027' \
-H 'accept: application/json' \
-H 'connection: close' \
-H 'X-Nick-Name: cccccccc-good-0000-0000-000000000000' \
-H 'x-auth-token: 2ca8a58ec971bfc6b0bcda8443931883eb5900d5' \
-H 'X-Display-Name: ' \
-H 'X-Roles: [{\"id\":\"security-role-0000-0000-000000000000\",\"name\":\"Security Team\"}]' \
-H 'X-Organizations: []' \
-H 'X-Eidas-Profile: {}' \
-H 'X-App-Id: tutorial-dckr-site-0000-xpresswebapp' \
-H 'x-dedwarfor-for: ::ffff:172.18.1.7' \
--data-raw '{
  "ring": {
      "type" : "command",
      "value" : ""
  }
}'
```

The result may bring new insights to the case.

**Comment by jason-fox** [comment]
-  `x-dedwarfor-for` rings the bell 🔔 
-  `x-forwarded-for` crashes the 3.1.0 Broker  🔕

**Comment by fgalan** [comment]
I have seen something interesting, looking traces carefully.

In your report:

```
time=2021-06-15T12:26:35.109Z | lvl=INFO | corr=ec9827d2-cdd4-11eb-8643-0242ac120109 | trans=1623759885-620-00000000026 | from=::ffff:127.0.0.1<none> | srv=<none> | subsrv=<none> | 
time=2021-06-15T12:27:34.211Z | lvl=ERROR | corr=0fd1c302-cdd5-11eb-8465-0242ac120109 | trans=1623759885-620-00000000039 | from=::ffff:172.18.1./ | srv=openiot | subsrv=/ | comp=Orion | op=postUpdateContext.cpp[642]:postUpdateContext | msg=Internal Error (attribute 'ring' not found)
time=2021-06-15T12:27:34.212Z | lvl=INFO | corr=0fd1c302-cdd5-11eb-8465-0242ac120109 | trans=1623759885-620-00000000039 | from=::ffff:172.18.1./ | srv=openiot | subsrv=/ | comp=Orion | op=logTracing.cpp[146]:logInfoFwdStart | msg=Starting forwarding for PATCH /v2/entities/Bell:001/attrs
time=2021-06-15T12:27:35.771Z | lvl=INFO | corr=10c0487e-cdd5-11eb-8120-0242ac120109 | trans=1623759885-620-00000000041 | from=172.18.1.9 | srv=<none> | subsrv=<none> | comp=Orion | op=logTracing.cpp[79]:logInfoRequestWithoutPayload | msg=Request received: GET /version, response code: 200

time=2021-06-15T12:38:19.987Z | lvl=INFO | corr=90bc074c-cdd6-11eb-a181-0242ac120109 | trans=1623760540-821-00000000016 | from=::ffff:127.0.0.1<none> | srv=<none> | subsrv=<none> | comp=Orion | op=logTracing.cpp[79]:logInfoRequestWithoutPayload | msg=Request received: GET /version, response code: 200
time=2021-06-15T12:38:22.328Z | lvl=ERROR | corr=922047ec-cdd6-11eb-abb9-0242ac120109 | trans=1623760540-821-00000000017 | from=::ffff:172.18.1./ | srv=openiot | subsrv=/ | comp=Orion | op=postUpdateContext.cpp[642]:postUpdateContext | msg=Internal Error (attribute 'ring' not found)
time=2021-06-15T12:38:22.328Z | lvl=INFO | corr=922047ec-cdd6-11eb-abb9-0242ac120109 | trans=1623760540-821-00000000017 | from=::ffff:172.18.1./ | srv=openiot | subsrv=/ | comp=Orion | op=logTracing.cpp[146]:logInfoFwdStart | msg=Starting forwarding for PATCH /v2/entities/Bell:001/attrs
```

Taking apart time=, lvl=, coor= and trans=

```
from=::ffff:127.0.0.1<none> | srv=<none> | subsrv=<none> | 
from=::ffff:172.18.1./ | srv=openiot | subsrv=/ | comp=Orion | op=postUpdateContext.cpp[642]:postUpdateContext | msg=Internal Error (attribute 'ring' not found)
from=::ffff:172.18.1./ | srv=openiot | subsrv=/ | comp=Orion | op=logTracing.cpp[146]:logInfoFwdStart | msg=Starting forwarding for PATCH /v2/entities/Bell:001/attrs
from=172.18.1.9 | srv=<none> | subsrv=<none> | comp=Orion | op=logTracing.cpp[79]:logInfoRequestWithoutPayload | msg=Request received: GET /version, response code: 200

from=::ffff:127.0.0.1<none> | srv=<none> | subsrv=<none> | comp=Orion | op=logTracing.cpp[79]:logInfoRequestWithoutPayload | msg=Request received: GET /version, response code: 200
from=::ffff:172.18.1./ | srv=openiot | subsrv=/ | comp=Orion | op=postUpdateContext.cpp[642]:postUpdateContext | msg=Internal Error (attribute 'ring' not found)
from=::ffff:172.18.1./ | srv=openiot | subsrv=/ | comp=Orion | op=logTracing.cpp[146]:logInfoFwdStart | msg=Starting forwarding for PATCH /v2/entities/Bell:001/attrs
```

We can see that:

* First trace doesn't follow the log format. It "truncates" after comp=. In addition, the from= field is weird, the token `<none>` should not be there
* We see also that unexpected `<none>` token in the GET /version request (5th trace)
* All the other traces have `from=::ffff:172.18.1./`. That's not correct, it should be `::ffff:172.18.1.7`

In the logs of my traces we see `::ffff:172.18.1.0.0.0.0`. In this case, Orion doesn't crash but I understand it should be `::ffff:172.18.1.7`

Maybe there is some issue in the log library with the from= field? Maybe it assumes IPv4-sized addressed and doesn't work correctly with IPv6 addresses, which are longer, causing some kind of buffer overflow problem in this case. @kzangeli (author of the log library :) what do you thing? @jason-fox could you do a test using a IPv4 address (e.g. -H 'x-forwarded-for: 172.18.1.7')?

**Comment by kzangeli** [comment]
I'll take a look

**Comment by kzangeli** [comment]
Running it under valgrind would give all the info, if it is a case of memory corruption, which is what I assume you are suspecting.

**Comment by jason-fox** [comment]
>  could you do a test using a IPv4 address (e.g. -H 'x-forwarded-for: 172.18.1.7')?

-  `x-forwarded-for: 172.18.1.7` - rings the bell 🔔 
-  `x-forwarded-for: ffff:172.18.1.7` - rings the bell 🔔 
-  `x-forwarded-for: ::ffff:172.18.1.7` - crashes the 3.1.0 Broker 🔕 

I guess the broker doesn't like colons.

**Comment by fgalan** [comment]
> I guess the broker doesn't like colons.

Second case has a colon in the middle. Maybe colons *as prefix* are the ones it doesn't like...

**Comment by jason-fox** [comment]
With more investigation, this appears to be an [Express IP4/IP6 issue](https://stackoverflow.com/questions/29411551/express-js-req-ip-is-returning-ffff127-0-0-1). I've just raised a PR to fix this specific issue in the code of the Wilma PEP-Proxy - https://github.com/ging/fiware-pep-proxy/pull/132 - however it still remains an issue for Orion because you should still put some defensive code in place (since you shouldn't be able crash the context broker with a weird `x-forwarded-for`). 

Effectively this is now a security issue for you since the flaw could be used for a denial-of-service attack.

**Comment by fgalan** [comment]
Proposed fix in PR https://github.com/telefonicaid/fiware-orion/pull/3917 (but probably it's not the smarter solution).

**Comment by fgalan** [comment]
Unfortunately, I didn't get feeback on this before closing release 3.2.0, so this issue is postponed to the next release (3.3.0).

**Comment by jason-fox** [comment]
Still an issue with 3.2.0

### Orion 3.0.0 and Wilma 8.1.0 - works

```text
type":"Text","value":"light"},"function":{"type":"Text","value":["onOff","sensing"]},"supportedProtocol":{"type":"Text","value":["ul20"]},"supportedUnits":{"type":"Text","value":"CDL"},"on_status":{"type":"commandStatus","value":"UNKNOWN"},"on_info":{"type":"commandResult","value":" "},"off_status":{"type":"commandStatus","value":"UNKNOWN"},"off_info":{"type":"commandResult","value":" "},"TimeInstant":{"type":"DateTime","value":"2021-09-15T14:07:02.110Z"}}, response code: 204
time=2021-09-15T14:07:12.924Z | lvl=INFO | corr=396aab10-162e-11ec-8ba4-0242ac120104 | trans=1631714786-575-00000000040 | from=172.18.1.4 | srv=<none> | subsrv=<none> | comp=Orion | op=logTracing.cpp[79]:logInfoRequestWithoutPayload | msg=Request received: GET /version, response code: 200
time=2021-09-15T14:07:25.879Z | lvl=INFO | corr=41221f28-162e-11ec-af08-0242ac120104 | trans=1631714786-575-00000000041 | from=::ffff:127.0.0.1<none> | srv=<none> | subsrv=<none> | comp=Orion | op=logTracing.cpp[79]:logInfoRequestWithoutPayload | msg=Request received: GET /version, response code: 200
time=2021-09-15T14:07:28.473Z | lvl=INFO | corr=42abb00c-162e-11ec-bdec-0242ac120104 | trans=1631714786-575-00000000042 | from=172.18.1.4 | srv=<none> | subsrv=<none> | comp=Orion | op=logTracing.cpp[79]:logInfoRequestWithoutPayload | msg=Request received: GET /version, response code: 200
time=2021-09-15T14:07:43.746Z | lvl=INFO | corr=4bc9a78e-162e-11ec-bf26-0242ac120104 | trans=1631714786-575-00000000043 | from=172.18.1.4 | srv=<none> | subsrv=<none> | comp=Orion | op=logTracing.cpp[79]:logInfoRequestWithoutPayload | msg=Request received: GET /version, response code: 200
time=2021-09-15T14:07:46.609Z | lvl=ERROR | corr=4d7d8de8-162e-11ec-af18-0242ac120104 | trans=1631714786-575-00000000044 | from=::ffff:172.18.1./ | srv=openiot | subsrv=/ | comp=Orion | op=postUpdateContext.cpp[632]:postUpdateContext | msg=Internal Error (attribute 'ring' not found)
time=2021-09-15T14:07:46.609Z | lvl=INFO | corr=4d7d8de8-162e-11ec-af18-0242ac120104 | trans=1631714786-575-00000000044 | from=::ffff:172.18.1./ | srv=openiot | subsrv=/ | comp=Orion | op=logTracing.cpp[146]:logInfoFwdStart | msg=Starting forwarding for PATCH /v2/entities/Bell:001/attrs
time=2021-09-15T14:07:46.688Z | lvl=INFO | corr=4d89cffe-162e-11ec-8f00-0242ac120104 | trans=1631714786-575-00000000046 | from=172.18.1.6 | srv=openiot | subsrv=/ | comp=Orion | op=logTracing.cpp[130]:logInfoRequestWithPayload | msg=Request received: POST /v2/entities/Bell:001/attrs?type=Bell, request payload (890 bytes): {"ring_status":{"type":"commandStatus","value":"PENDING","metadata":{"TimeInstant":{"type":"DateTime","value":"2021-09-15T14:07:46.661Z"}}},"refStore":{"type":"Relationship","value":"urn:ngsi-ld:Store:001","metadata":{"TimeInstant":{"type":"DateTime","value":"2021-09-15T14:07:46.661Z"}}},"category":{"type":"Text","value":["actuator"],"metadata":{"TimeInstant":{"type":"DateTime","value":"2021-09-15T14:07:46.661Z"}}},"controlledProperty":{"type":"Text","value":"noiseLevel","metadata":{"TimeInstant":{"type":"DateTime","value":"2021-09-15T14:07:46.661Z"}}},"function":{"type":"Text","value":["onOff"],"metadata":{"TimeInstant":{"type":"DateTime","value":"2021-09-15T14:07:46.661Z"}}},"supportedProtocol":{"type":"Text","value":["ul20"],"metadata":{"TimeInstant":{"type":"DateTime","value":"2021-09-15T14:07:46.661Z"}}},"TimeInstant":{"type":"DateTime","value":"2021-09-15T14:07:46.661Z"}}, response code: 204
time=2021-09-15T14:07:46.700Z | lvl=INFO | corr=4d7d8de8-162e-11ec-af18-0242ac120104; cbfwd=1 | trans=1631714786-575-00000000045 | from=::ffff:172.18.1./ | srv=openiot | subsrv=/ | comp=Orion | op=logTracing.cpp[212]:logInfoFwdRequest | msg=Request forwarded (regId: 6141fe0493cfc604b87ac42f): POST http://iot-agent:4041/op/update, request payload (119 bytes): {"entities":[{"id":"Bell:001","type":"Bell","ring":{"type":"command","value":"","metadata":{}}}],"actionType":"update"}, response payload (0 bytes): , response code: 204
time=2021-09-15T14:07:46.702Z | lvl=INFO | corr=4d7d8de8-162e-11ec-af18-0242ac120104 | trans=1631714786-575-00000000044 | from=::ffff:172.18.1./ | srv=openiot | subsrv=/ | comp=Orion | op=logTracing.cpp[130]:logInfoRequestWithPayload | msg=Request received: PATCH /v2/entities/Bell:001/attrs, request payload (38 bytes): {"ring":{"type":"command","value":""}}, response code: 204
time=2021-09-15T14:07:46.753Z | lvl=INFO | corr=4d93bbe0-162e-11ec-9327-0242ac120104 | trans=1631714786-575-00000000047 | from=172.18.1.6 | srv=openiot | subsrv=/ | comp=Orion | op=logTracing.cpp[130]:logInfoRequestWithPayload | msg=Request received: POST /v2/entities/Bell:001/attrs?type=Bell, request payload (1023 bytes): {"ring_status":{"type":"commandStatus","value":"OK","metadata":{"TimeInstant":{"type":"DateTime","value":"2021-09-15T14:07:46.739Z"}}},"ring_info":{"type":"commandResult","value":" ring OK","metadata":{"TimeInstant":{"type":"DateTime","value":"2021-09-15T14:07:46.739Z"}}},"refStore":{"type":"Relationship","value":"urn:ngsi-ld:Store:001","metadata":{"TimeInstant":{"type":"DateTime","value":"2021-09-15T14:07:46.739Z"}}},"category":{"type":"Text","value":["actuator"],"metadata":{"TimeInstant":{"type":"DateTime","value":"2021-09-15T14:07:46.739Z"}}},"controlledProperty":{"type":"Text","value":"noiseLevel","metadata":{"TimeInstant":{"type":"DateTime","value":"2021-09-15T14:07:46.739Z"}}},"function":{"type":"Text","value":["onOff"],"metadata":{"TimeInstant":{"type":"DateTime","value":"2021-09-15T14:07:46.739Z"}}},"supportedProtocol":{"type":"Text","value":["ul20"],"metadata":{"TimeInstant":{"type":"DateTime","value":"2021-09-15T14:07:46.739Z"}}},"TimeInstant":{"type":"DateTime","value":"2021-09-15T14:07:46.739Z"}}, response code: 204
time=2021-09-15T14:07:58.916Z | lvl=INFO | corr=54d45cac-162e-11ec-b952-0242ac120104 | trans=1631714786-575-00000000048 | from=172.18.1.4 | srv=<none> | subsrv=<none> | comp=Orion | op=logTracing.cpp[79]:logInfoRequestWithoutPayload | msg=Request received: GET /version, response code: 200
time=2021-09-15T14:08:14.297Z | lvl=INFO | corr=5dff6164-162e-11ec-9a24-0242ac120104 | trans=1631714786-575-00000000049 | from=172.18.1.4 | srv=<none> | subsrv=<none> | comp=Orion | op=logTracing.cpp[79]:logInfoRequestWithoutPayload | msg=Request received: GET /version, response code: 200
time=2021-09-15T14:08:16.258Z | lvl=INFO | corr=5f2a972a-162e-11ec-b536-0242ac120104 | trans=1631714786-575-00000000050 | from=::ffff:127.0.0.1<none> | srv=<none> | subsrv=<none> | comp=Orion | op=logTracing.cpp[79]:logInfoRequestWithoutPayload | msg=Request received: GET /version, response code: 200
```

### Orion 3.2.0 and Wilma 8.1.0 - crashes

```text
time=2021-09-15T14:01:45.517Z | lvl=INFO | corr=764354ac-162d-11ec-9ff0-0242ac120104 | trans=1631714434-042-00000000048 | from=::ffff:172.18.1.<none> | srv=<none> | subsrv=<none> | comp=Orion | op=logTracing.cpp[90]:logInfoRequestWithoutPayload | msg=Request received: GET /v2/entities/urn:ngsi-ld:InventoryItem:001?options=keyValues&type=InventoryItem, response code: 200
time=2021-09-15T14:01:45.685Z | lvl=INFO | corr=765c7c3e-162d-11ec-96f6-0242ac120104 | trans=1631714434-042-00000000049 | from=::ffff:172.18.1.<none> | srv=<none> | subsrv=<none> | comp=Orion | op=logTracing.cpp[141]:logInfoRequestWithPayload | msg=Request received: PATCH /v2/entities/urn:ngsi-ld:InventoryItem:001/attrs, request payload (44 bytes): {"shelfCount":{"type":"Integer","value":14}}, response code: 204
time=2021-09-15T14:01:45.896Z | lvl=INFO | corr=767da8b4-162d-11ec-8860-0242ac120104 | trans=1631714434-042-00000000050 | from=::ffff:172.18.1.<none> | srv=<none> | subsrv=<none> | comp=Orion | op=logTracing.cpp[90]:logInfoRequestWithoutPayload | msg=Request received: GET /v2/entities?options=keyValues&type=Product, response code: 200
time=2021-09-15T14:01:45.900Z | lvl=INFO | corr=767dfefe-162d-11ec-85f6-0242ac120104 | trans=1631714434-042-00000000051 | from=::ffff:172.18.1.<none> | srv=<none> | subsrv=<none> | comp=Orion | op=logTracing.cpp[90]:logInfoRequestWithoutPayload | msg=Request received: GET /v2/entities?q=refStore%3D%3Durn%3Angsi-ld%3AStore%3A001&options=keyValues&type=InventoryItem, response code: 200
time=2021-09-15T14:01:50.641Z | lvl=INFO | corr=7950326e-162d-11ec-a6d0-0242ac120104 | trans=1631714434-042-00000000052 | from=172.18.1.4 | srv=<none> | subsrv=<none> | comp=Orion | op=logTracing.cpp[90]:logInfoRequestWithoutPayload | msg=Request received: GET /version, response code: 200
time=2021-09-15T14:01:54.046Z | lvl=WARN | corr=7b594672-162d-11ec-b3d5-0242ac120104 | trans=1631714434-042-00000000053 | from=::ffff:172.18.1./ | srv=openiot | subsrv=/ | comp=Orion | op=AlarmManager.cpp[438]:badInput | msg=Raising alarm BadInput 172.18.1.7: attribute not exists
time=2021-09-15T14:01:54.048Z | lvl=ERROR | corr=7b594672-162d-11ec-b3d5-0242ac120104 | trans=1631714434-042-00000000053 | from=::ffff:172.18.1./ | srv=openiot | subsrv=/ | comp=Orion | op=postUpdateContext.cpp[643]:postUpdateContext | msg=Internal Error (attribute 'ring' not found)
time=2021-09-15T14:01:54.049Z | lvl=ERROR | corr=7b594672-162d-11ec-b3d5-0242ac120104 | trans=1631714434-042-00000000053 | from=::ffff:172.18.1./ | srv=openiot | subsrv=/ | comp=Orion | op=postUpdateContext.cpp[682]:postUpdateContext | msg=Orion Bug (empty contextAttributeVector for ContextElementResponse 1)
time=2021-09-15T14:01:54.049Z | lvl=INFO | corr=7b594672-162d-11ec-b3d5-0242ac120104 | trans=1631714434-042-00000000053 | from=::ffff:172.18.1./ | srv=openiot | subsrv=/ | comp=Orion | op=logTracing.cpp[157]:logInfoFwdStart | msg=Starting forwarding for PATCH /v2/entities/Bell:001/attrs
free(): invalid pointer
```

**Comment by fgalan** [comment]
> Still an issue with 3.2.0

Makes sense. As I told above:

> Unfortunately, I didn't get feeback on this before closing release 3.2.0, so this issue is postponed to the next release (3.3.0).

**Comment by fgalan** [comment]
> @jason-fox could you apply the same change locally to your CB and test again to see if that solves the issue, please?

I have built an image in dockerhub with the fix in PR https://github.com/telefonicaid/fiware-orion/pull/3917. This one: **telefonicaiot/fiware-orion:issue3884**

**Comment by jason-fox** [comment]
That build works fine - log below:


```text
time=2021-09-15T15:10:28.867Z | lvl=INFO | corr=0ff7d74a-1637-11ec-a4a1-0242ac120104 | trans=1631718549-106-00000000055 | from=172.18.1.6 | srv=openiot | subsrv=/ | comp=Orion | op=logTracing.cpp[130]:logInfoRequestWithPayload | msg=Request received: POST /v2/op/update, request payload (1090 bytes): {"actionType":"update","entities":[{"id":"Bell:001","type":"Bell","ring_status":{"type":"commandStatus","value":"OK","metadata":{"TimeInstant":{"type":"DateTime","value":"2021-09-15T15:10:28.846Z"}}},"ring_info":{"type":"commandResult","value":" ring OK","metadata":{"TimeInstant":{"type":"DateTime","value":"2021-09-15T15:10:28.846Z"}}},"refStore":{"type":"Relationship","value":"urn:ngsi-ld:Store:001","metadata":{"TimeInstant":{"type":"DateTime","value":"2021-09-15T15:10:28.846Z"}}},"category":{"type":"Text","value":["actuator"],"metadata":{"TimeInstant":{"type":"DateTime","value":"2021-09-15T15:10:28.846Z"}}},"controlledProperty":{"type":"Text","value":"noiseLevel","metadata":{"TimeInstant":{"type":"DateTime","value":"2021-09-15T15:10:28.846Z"}}},"function":{"type":"Text","value":["onOff"],"metadata":{"TimeInstant":{"type":"DateTime","value":"2021-09-15T15:10:28.846Z"}}},"supportedProtocol":{"type":"Text","value":["ul20"],"metadata":{"TimeInstant":{"type":"DateTime","value":"2021-09-15T15:10:28.846Z"}}},"TimeInstant":{"type":"DateTime","value":"2021-09-15T15:10:28.846Z"}}]}, response code: 204
time=2021-09-15T15:10:42.524Z | lvl=INFO | corr=181b0bb8-1637-11ec-972b-0242ac120104 | trans=1631718549-106-00000000056 | from=172.18.1.4 | srv=<none> | subsrv=<none> | comp=Orion | op=logTracing.cpp[79]:logInfoRequestWithoutPayload | msg=Request received: GET /version, response code: 200
time=2021-09-15T15:10:57.936Z | lvl=INFO | corr=214bbd90-1637-11ec-a57d-0242ac120104 | trans=1631718549-106-00000000057 | from=172.18.1.4 | srv=<none> | subsrv=<none> | comp=Orion | op=logTracing.cpp[79]:logInfoRequestWithoutPayload | msg=Request received: GET /version, response code: 200
```

**Comment by fgalan** [comment]
Thanks for the report!

However, not sure if the proposed fix is enough. I have just extended the length from 15 to 45 to cope with IPv6 addresses in "numeric" format. However, I guess that something like:

```
curl ... -H 'x-forwarded-for: a-very-looooooooooooooooooooooooooooong-string-greater-than-45-chars' ...
```

is still problematic.

It would be great if the author of the log library @kzangeli could have a look to this and provide a fix to cover also that situations.

**Comment by kzangeli** [comment]
Sorry, I promised to have a look. Been out for parental leave (16 weeks) and I forgot about the whole thing.
I'll try to look at this *this* week.

Do we have an easy way to reproduce the issue?

**Comment by mapedraza** [comment]
Thank for your willingness to solve this bug

When Orion receive a message with the `x-forwarded-for` header longer than expected, it crashes, as described by @jason-fox 

>
>-  `x-forwarded-for: 172.18.1.7` - rings the bell 🔔 
>-  `x-forwarded-for: ffff:172.18.1.7` - rings the bell 🔔 
>-   `x-forwarded-for: :ffff:172.18.1.7` - crashes the 3.1.0 Broker 🔕 
> -  `x-forwarded-for: ::ffff:172.18.1.7` - crashes the 3.1.0 Broker 🔕 

As @fgalan analyzed in the code, it comes from `logMsq.cpp`

>Looking from `FROM_IP` in logMsq.cpp:
>
>```
>   else if (strncmp(&format[fi], "FROM_IP", 7) == 0)
>    {
>      STRING_ADD(fromIp, 7);
>    }
>```
>
>Looking on how that `fromIp` var is defined (logMsq.h):
>
>```
>extern __thread char   fromIp[IP_LENGTH_MAX + 1];
>```
>
>and finally in limits.h
>
>```
>#define IP_LENGTH_MAX           15     // Based on xxx.xxx.xxx.xxx
>```

Changing `IP_LENGTH_MAX` make it works but It would still problematic when making a request longer than spected

**Comment by fgalan** [comment]
Finally fixed by PR #4039 (== PR #4026)
